ext/winevt/winevt_locale.c

```diff
@@ -0,0 +1,69 @@
+    { MAKELANGID(LANG_ENGLISH,    SUBLANG_ENGLISH_CAN),          "en_CA"},
+    { MAKELANGID(LANG_ENGLISH,    SUBLANG_ENGLISH_NZ),           "en_NZ"},
+    { MAKELANGID(LANG_ENGLISH,    SUBLANG_ENGLISH_EIRE),         "en_IE"},
+    { MAKELANGID(LANG_FINNISH,    SUBLANG_DEFAULT),              "fi_FI"},
+    { MAKELANGID(LANG_FRENCH,     SUBLANG_FRENCH),               "fr_FR"},
+    { MAKELANGID(LANG_FRENCH,     SUBLANG_FRENCH_BELGIAN),       "fr_BE"},
+    { MAKELANGID(LANG_FRENCH,     SUBLANG_FRENCH_CANADIAN),      "fr_CA"},
+    { MAKELANGID(LANG_FRENCH,     SUBLANG_FRENCH_SWISS),         "fr_CH"},
+    { MAKELANGID(LANG_GERMAN,     SUBLANG_GERMAN),               "de_DE"},
+    { MAKELANGID(LANG_GERMAN,     SUBLANG_GERMAN_SWISS),         "de_CH"},
+    { MAKELANGID(LANG_GERMAN,     SUBLANG_GERMAN_AUSTRIAN),      "de_AT"},
+    { MAKELANGID(LANG_GREEK,      SUBLANG_DEFAULT),              "el_GR"},
+    { MAKELANGID(LANG_HUNGARIAN,  SUBLANG_DEFAULT),              "hu_HU"},
+    { MAKELANGID(LANG_ICELANDIC,  SUBLANG_DEFAULT),              "is_IS"},
+    { MAKELANGID(LANG_ITALIAN,    SUBLANG_ITALIAN),              "it_IT"},
+    { MAKELANGID(LANG_ITALIAN,    SUBLANG_ITALIAN_SWISS),        "it_CH"},
+    { MAKELANGID(LANG_JAPANESE,   SUBLANG_DEFAULT),              "ja_JP"},
+    { MAKELANGID(LANG_KOREAN,     SUBLANG_DEFAULT),              "ko_KO"},
+    { MAKELANGID(LANG_NORWEGIAN,  SUBLANG_NORWEGIAN_BOKMAL),     "no_NO"},
+    { MAKELANGID(LANG_NORWEGIAN,  SUBLANG_NORWEGIAN_BOKMAL),     "nb_NO"},
+    { MAKELANGID(LANG_NORWEGIAN,  SUBLANG_NORWEGIAN_NYNORSK),    "nn_NO"},
+    { MAKELANGID(LANG_POLISH,     SUBLANG_DEFAULT),              "pl_PL"},
+    { MAKELANGID(LANG_PORTUGUESE, SUBLANG_PORTUGUESE),           "pt_PT"},
+    { MAKELANGID(LANG_PORTUGUESE, SUBLANG_PORTUGUESE_BRAZILIAN), "pt_BR"},
+    { MAKELANGID(LANG_ROMANIAN,   SUBLANG_DEFAULT),              "ro_RO"},
+    { MAKELANGID(LANG_RUSSIAN,    SUBLANG_DEFAULT),              "ru_RU"},
+    { MAKELANGID(LANG_SLOVAK,     SUBLANG_DEFAULT),              "sk_SK"},
+    { MAKELANGID(LANG_SLOVENIAN,  SUBLANG_DEFAULT),              "sl_SI"},
+    { MAKELANGID(LANG_SPANISH,    SUBLANG_SPANISH),              "es_ES"},
+    { MAKELANGID(LANG_SPANISH,    SUBLANG_SPANISH),              "es_ES_T"},
+    { MAKELANGID(LANG_SPANISH,    SUBLANG_SPANISH_MEXICAN),      "es_MX"},
+    { MAKELANGID(LANG_SPANISH,    SUBLANG_SPANISH_MODERN),       "es_ES_M"},
+    { MAKELANGID(LANG_SWEDISH,    SUBLANG_DEFAULT),              "sv_SE"},
+    { MAKELANGID(LANG_TURKISH,    SUBLANG_DEFAULT),              "tr_TR"},
+    { 0, NULL}
+};
+
+const LocaleInfo default_locale = {MAKELANGID(LANG_NEUTRAL, SUBLANG_NEUTRAL), "neutral"};
+
+LocaleInfo
+get_locale_from_rb_str(VALUE rb_locale_str)
+{
+  CHAR* locale_str = StringValuePtr(rb_locale_str);
+  int localeInfoTableSize = _countof(localeInfoTable);
+
+  for (int i = 0; i < localeInfoTableSize; i++) {
+    if (stricmp(localeInfoTable[i].langCode, locale_str) == 0) {
```

Comment:
It will cause crash when no `LocaleInfo` is matched since last `langCode` is `NULL` and it will be passed to `stricmp`.

Reply:
Last langCode is not used because this should be used for just calculating table upper bound.
 
```c
for (int i = 0; i <= localeInfoTableSize; i++) {
```

should cause NULL pointer error but the current code does not include the last element as upper bound.
Am I wrong?

Reply:
If the array is

```c
static LocaleInfo localeInfoTable [] = {
    { MAKELANGID(LANG_BULGARIAN,  SUBLANG_DEFAULT),              "bg_BG"},
    { 0, NULL}
};
```

`localeInfoTableSize` is `2`, so that `i` can be `0` and `1` for `i < localeInfoTableSize`.
`localeInfoTable[1]` points `{ 0, NULL}`.

Since this array has a terminator element, removing `localeInfoTableSize` and using the following condition is simpler than it, it's a popular idiom in C lang:

```c
 for (int i = 0; localeInfoTableSize[i].langCode != NULL; i++) {
```

Reply:
Fixed in https://github.com/fluent-plugins-nursery/winevt_c/pull/27/commits/88c55afc22aa6399a5b7cee9703a682eab419d66.